### PR TITLE
Add test mode

### PR DESCRIPTION
**Pull request by yellowjaguar5**
Ported from #29 with minor edits. Closes #29. 
One question: should the documentation on the Sendable interface use Shuffleboard methods instead of SmartDashboard methods?

Edit: **I'll be away from the internet for two weeks (back on August 11th) and I plan to continue working on this PR when I return.**

### CODE REVIEW

File: source/docs/software/networktables/creating-sendable-classes.rst [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+                builder.addDoubleProperty("Angle", () -> armSubsystem::getAngle, (value) -> armSubsystem:setAngle(value));
+                builder.addBooleanProperty("IsOpen", () -> armSubsystem::isOpen, (value) -> armSubsystem:setClaw(value));
+
+                super.initSendable(builder);
+            }
+        }
+
+  .. code-tab:: c++
+
+    Todo: add C++ code
```
Comment (yellowjaguar5) [inline source/docs/software/networktables/creating-sendable-classes.rst]: Could someone suggest a C++ equivalent example?

**Comment by yellowjaguar5** [comment]
Also, could someone suggest a C++ example for creating a new Sendable class with SendableBuilder?

File: source/docs/software/networktables/creating-sendable-classes.rst [hunk 2]
```diff
@@ -0,0 +1,160 @@
+.. _creating-sendable-classes:
```
Comment (Daltz333) [inline source/docs/software/networktables/creating-sendable-classes.rst]: Remove relative references. Use absolute references only.

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+Enabling Test mode
+------------------
+
+Test mode on the robot can be enabled from the Driver Station just like Autonomous or Teleop. When in test mode, the testInit method is run once, and the testPeriodic method is run once per tick, in addition to RobotPeriodic, the same as Teleop and Autonomous control modes. To enable test mode in the Driver Station, select the "Test" button and enable the robot. The test mode code will then run.
+
+Adding Test mode code to your robot code
+----------------------------------------
+
+Adding test mode can be as painless as calling your already written Teleop methods from Test. This will allow you to use the LiveWindow tuning features for classes such as PIDBase and PIDController, as well as PIDSubsystems and PIDCommands, to change PIDF constants and setpoints. Make sure to add your subsystems to SmartDashboard with the .. code-block:: putData(subsystem) and .. code-block:: PutData(subsystem) methods.
```
Comment (Daltz333) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
Adding test mode can be as painless as calling your already written Teleop methods from Test. This will allow you to use the LiveWindow tuning features for classes such as PIDBase and PIDController, as well as PIDSubsystems and PIDCommands, to change PIDF constants and setpoints. Make sure to add your subsystems to SmartDashboard with the ``putData(subsystem)`` and ``PutData(subsystem)`` methods.
```

File: source/docs/software/actuators/wpi-drive-classes.rst [hunk 4]
```diff
@@ -166,6 +166,8 @@ rule, so clockwise rotation around the Z axis is positive.
 
 .. warning:: This convention is different than the convention for joysticks which typically have -Y as Up (commonly mapped to throttle) and +X as Right. Pay close attention to the examples below if you want help with typical Joystick->Drive mapping.
 
+.. _wpi_differential_drive:
```
Comment (Daltz333) [inline source/docs/software/actuators/wpi-drive-classes.rst]: Remove relative reference

**Comment by yellowjaguar5** [comment]
I think this was ready to be merged, why was it closed?

**Comment by AustinShalit** [comment]
This was a mistake on my part... I deleted the develop branch after merging it into master and GitHub automatically closed all of the PRs

File: source/docs/software/networktables/creating-sendable-classes.rst [hunk 5]
```diff
@@ -0,0 +1,158 @@
+Using the Sendable interface
+============================
+
+What is Sendable?
+-----------------
+
+Sendable is the base class that allows something to be sent over NetworkTables for debugging through programs such as OutlineViewer, Shuffleboard and SmartDashboard. Base WPI classes such as Command and Subsystem, as well as actuator classes such as DoubleSolenoid, implement the Sendable interface. The WPI documentation of the Sendable interface, as well as a full list of WPI classes which implement Sendable, can be found on the WPI docs for Sendable (`Java <https://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/Sendable.html>`__, `C++ <https://first.wpi.edu/FRC/roborio/release/docs/cpp/classSendable.html>`__).
```
Comment (AustinShalit) [inline source/docs/software/networktables/creating-sendable-classes.rst]: Oxford comma. Slight wording change.

```suggestion
Sendable is the base class that allows something to be sent over NetworkTables for debugging through programs such as OutlineViewer, Shuffleboard, and SmartDashboard. Base WPI classes such as Command and Subsystem, as well as actuator classes such as DoubleSolenoid, implement the Sendable interface. The WPI documentation of the Sendable interface, as well as a full list of WPI classes which implement Sendable, can be found in the API docs for Sendable (`Java <https://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/Sendable.html>`__, `C++ <https://first.wpi.edu/FRC/roborio/release/docs/cpp/classSendable.html>`__).
```

File: source/docs/software/networktables/creating-sendable-classes.rst [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      frc::SmartDashboard::PutData("my sendable", m_SendableClass)
+
+      frc::ShuffleboardTab& m_tab = frc::Shuffleboard::GetTab("tabName");
+      m_tab.Add("Motor", m_Motor);
+
+
+Interacting with Sendable classes
+---------------------------------
+
+Sendable classes are published via NetworkTables for users to view using WPILIb through programs such as OutlineViewer, Shuffleboard and SmartDashboard. These programs allow users to not only view the current state, such as solenoid states, motor output, or PIDSubsystem/PIDCommand's current PID constants, but to actually modify or send data the other direction to the RoboRIO to do things like manually start and stop commands.
```
Comment (AustinShalit) [inline source/docs/software/networktables/creating-sendable-classes.rst]:
```suggestion
Sendable classes are published via NetworkTables for users to view using WPILIb through programs such as OutlineViewer, Shuffleboard, and SmartDashboard. These programs allow users to not only view the current state, such as solenoid states, motor output, or PIDSubsystem/PIDCommand's current PID constants, but to actually modify or send data the other direction to the RoboRIO to do things like manually start and stop commands.
```

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+Notice that the values are grouped by the subsystem names to group related actuators and sensors for easy testing. The subsystem names are suppled in the AddActuator() and AddSensor() method calls as shown in the code examples. This grouping, while not required, makes it much easier to test one subsystem at a time and have all the values next to each other on the screen. Using Test mode with the IterativeRobot template Using Test mode with the IterativeRobot template
+
+The IterativeRobot template lends itself quite nicely to testing since it will periodically call the testPeriodic() method (or Test() in C++) in your robot program. The testPeriodic() method will be called with each DriverStation update, about every 20ms, and it is a good place to do whatever testing commands or LiveWindow updates are desired for testing. The LiveWindow updating is built into the IterativeRobot template so there is very little that is necessary to do in the program to get LiveWindow updates.
```
Comment (Daltz333) [inline source/docs/software/wpilib-overview/using-test-mode.rst]: Replace IterativeRobot with TimedRobot

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+.. note:: This works even if you are using the IterativeRobot template and not doing Command-based programming.
```
Comment (Daltz333) [inline source/docs/software/wpilib-overview/using-test-mode.rst]: Use TimedRobot

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+In this example the sensors are registered with the LiveWindow and during the testPeriodic method, simply update all the values by calling the LiveWindow run() method. If your program is causing too much network traffic you can call the run method less frequently by, for example, only calling it every 5 updates for a 100ms update rate.
```
Comment (Daltz333) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
In this example the sensors are registered with the LiveWindow and during the testPeriodic method, simply update all the values by calling the LiveWindow ``run()`` method. If your program is causing too much network traffic you can call the run method less frequently by, for example, only calling it every 5 updates for a 100ms update rate.
```

**Comment by yellowjaguar5** [comment]
@AustinShalit Any more changes or is it ready to merge?

**Comment by Daltz333** [comment]
@SamCarlberg Mind reviewing the accuracy of the information?

**Comment by AustinShalit** [comment]
Needs a rebase.

**Comment by yellowjaguar5** [comment]
@AustinShalit please review

**Comment by Daltz333** [comment]
Sorry about the delay on this PR. There is a lot of rush handling pull requests in the main allwpilib repo due to an upcoming deadline.

This pull-request requires a bit more review on the authenticity by the creators of the sendable. The sendable interface is also getting reworked a bit. Sorry about this.

**Comment by yellowjaguar5** [comment]
You're all good @Daltz333. Thanks for the update.

File: source/index.rst [hunk 10]
```diff
@@ -62,8 +62,8 @@ New teams may wish to familiarize themselves with the different components of th
    docs/networking/networking-basics
    docs/networking/ip-configurations
    docs/networking/roborio-network-troubleshooting
-   docs/networking/windows-firewall-configuration
    docs/networking/measuring-bandwidth-usage
+   docs/networking/windows-firewall-configuration
```
Comment (AustinShalit) [inline source/index.rst]: Seems like an unnecessary change.

File: source/docs/software/networktables/creating-sendable-classes.rst (hunk 1 above)
Comment (AustinShalit) [reply source/docs/software/networktables/creating-sendable-classes.rst]: Still need an example here

File: source/docs/software/networktables/creating-sendable-classes.rst [hunk 11]
```diff
@@ -121,7 +121,24 @@ Creating an entirely new :code:`Sendable` class follows a similar pattern to the
...
+
+            void initSendable(SendableBuilder s) {
+              /*
+              builder.setSmartDashboardType("ArmState");
+              builder.addDoubleProperty("Angle", () -> armSubsystem::getAngle, (value) -> armSubsystem:setAngle(value));
+              builder.addBooleanProperty("IsOpen", () -> armSubsystem::isOpen, (value) -> armSubsystem:setClaw(value));
+              */
+              Sendable::initSendable(s);
+            }
+        }
```
Comment (yellowjaguar5) [inline source/docs/software/networktables/creating-sendable-classes.rst]: @Oblarg Can you help me fix this example so it makes more sense (and works)? I'm not well versed in object-oriented c++.

**Comment by yellowjaguar5** [comment]
I was just reading over the [New for 2020](https://docs.wpilib.org/en/latest/docs/software/wpilib-overview/new-for-2020.html#wpilib) article and noticed that the Sendable framework has changed. Should I remove the changes to Sendable in this PR and open an issue to add details on the new Sendable usage (since I'm not sure what they are)?

**Comment by AustinShalit** [comment]
Let's do that. Don't want to have old information in the docs.

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+Test mode in the SmartDashboard
+-------------------------------
+
+The above sample code produces the following output when the Driver Station is put into Test mode then enabled. You can operate the motors by moving the sliders and read the values of sensors such as the wrist potentiometer.
+
+Notice that the values are grouped by the subsystem names to group related actuators and sensors for easy testing. The subsystem names are suppled in the AddActuator() and AddSensor() method calls as shown in the code examples. This grouping, while not required, makes it much easier to test one subsystem at a time and have all the values next to each other on the screen.
+
+Using Test mode with the TimedRobot template
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (AustinShalit) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+The TimedRobot template lends itself quite nicely to testing since it will periodically call the testPeriodic() method (or Test() in C++) in your robot program. The testPeriodic() method will be called with each DriverStation update, about every 20ms, and it is a good place to do whatever testing commands or LiveWindow updates are desired for testing. The LiveWindow updating is built into the TimedRobot template so there is very little that is necessary to do in the program to get LiveWindow updates.
```
Comment (AustinShalit) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
The TimedRobot template lends itself quite nicely to testing since it will periodically call the testPeriodic() method (or TestPeriodic() in C++) in your robot program. The testPeriodic() method will be called every 20ms and it is a good place to test commands or have LiveWindow update. The LiveWindow updating is built into the TimedRobot template so there is very little work to use LiveWindow.
```

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+Enabling Test mode
+------------------
+
+Test mode on the robot can be enabled from the Driver Station just like Autonomous or Teleop. When in test mode, the testInit method is run once, and the testPeriodic method is run once per tick, in addition to RobotPeriodic, the same as Teleop and Autonomous control modes. To enable test mode in the Driver Station, select the "Test" button and enable the robot. The test mode code will then run.
+
+Adding Test mode code to your robot code
+----------------------------------------
+
+Adding test mode can be as painless as calling your already written Teleop methods from Test. This will allow you to use the LiveWindow tuning features for classes such as PIDBase and PIDController, as well as PIDSubsystems and PIDCommands, to change PIDF constants and setpoints. Make sure to add your subsystems to SmartDashboard with the ``putData(subsystem)`` and ``PutData(subsystem)`` methods.
```
Comment (AustinShalit) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
Adding test mode can be as painless as calling your already written Teleop methods from Test. This will allow you to use the LiveWindow tuning features for classes such as PIDBase and PIDController, as well as PIDSubsystems and PIDCommands, to change PID constants and setpoints. Make sure to add your subsystems to SmartDashboard with the ``putData(subsystem)`` or ``PutData(subsystem)`` method.
```

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 15]
```diff
@@ -0,0 +1,71 @@
+Using test mode
+===============
+
+Test mode is designed to enable programmers to have a place to put code to verify that all systems on the robot are functioning. In each of the robot program templates there is a place to add test code to the robot. If you use the TimedRobot template, or use command-based programming you can easily see the operation of all the motors and sensors through the LiveWindow.
+
+Enabling Test mode
+------------------
+
+Test mode on the robot can be enabled from the Driver Station just like Autonomous or Teleop. When in test mode, the testInit method is run once, and the testPeriodic method is run once per tick, in addition to RobotPeriodic, the same as Teleop and Autonomous control modes. To enable test mode in the Driver Station, select the "Test" button and enable the robot. The test mode code will then run.
```
Comment (AustinShalit) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
Test mode on the robot can be enabled from the Driver Station just like autonomous or teleop. When in test mode, the ``testInit`` method is run once, and the testPeriodic method is run once per tick, in addition to ``robotPeriodic``, the same as teleop and autonomous control modes. To enable test mode in the Driver Station, select the "Test" button and enable the robot. The test mode code will then run.
```

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+        void Robot::TestInit() {
+            TestInit();
+        }
+
+        void Robot::TestPeriodic() {
+            TeleopPeriodic();
+        }
+
+Test mode in the SmartDashboard
```
Comment (Daltz333) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
Test Mode in SmartDashboard
```

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+-------------------------------
+
+The above sample code produces the following output when the Driver Station is put into Test mode then enabled. You can operate the motors by moving the sliders and read the values of sensors such as the wrist potentiometer.
+
+Notice that the values are grouped by the subsystem names to group related actuators and sensors for easy testing. The subsystem names are suppled in the AddActuator() and AddSensor() method calls as shown in the code examples. This grouping, while not required, makes it much easier to test one subsystem at a time and have all the values next to each other on the screen.
+
+Using Test mode with the TimedRobot template
```
Comment (Daltz333) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
Using Test Mode with the TimedRobot Template
```

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 18]
```diff
@@ -0,0 +1,71 @@
+Using test mode
```
Comment (Daltz333) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
Using Test Mode
```

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+===============
+
+Test mode is designed to enable programmers to have a place to put code to verify that all systems on the robot are functioning. In each of the robot program templates there is a place to add test code to the robot. If you use the TimedRobot template, or use command-based programming you can easily see the operation of all the motors and sensors through the LiveWindow.
+
+Enabling Test mode
```
Comment (Daltz333) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
Enabling Test Mode
```

File: source/docs/software/wpilib-overview/using-test-mode.rst [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+Using Test mode with the TimedRobot template
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The TimedRobot template lends itself quite nicely to testing since it will periodically call the testPeriodic() method (or Test() in C++) in your robot program. The testPeriodic() method will be called with each DriverStation update, about every 20ms, and it is a good place to do whatever testing commands or LiveWindow updates are desired for testing. The LiveWindow updating is built into the TimedRobot template so there is very little that is necessary to do in the program to get LiveWindow updates.
+
+.. note:: This works even if you are using the TimedRobot template and not doing Command-based programming.
+
+In this example the sensors are registered with the LiveWindow and during the testPeriodic method, simply update all the values by calling the LiveWindow ``run()`` method. If your program is causing too much network traffic you can call the run method less frequently by, for example, only calling it every 5 updates for a 100ms update rate.
+
+PID Tuning in Test mode
```
Comment (Daltz333) [inline source/docs/software/wpilib-overview/using-test-mode.rst]:
```suggestion
PID Tuning in Test Mode
```
